src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,332 @@
+/**
+ * @file gan.hpp
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_IMPL_HPP
+#define MLPACK_METHODS_ANN_GAN_IMPL_HPP
+
+#include "gan.hpp"
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+
+#include <mlpack/methods/ann/activation_functions/softplus_function.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/core/dists/gaussian_distribution.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::math;
+using namespace mlpack::distribution;
+
+namespace mlpack {
+namespace ann /** artifical neural network  */ {
+template<typename Generator, typename Discriminator, typename IntializerType>
+GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::GenerativeAdversarialNetwork(arma::mat& trainData,
+    IntializerType initializeRule,
+    Generator& generator,
+    Discriminator& discriminator,
+    size_t batchSize,
+    size_t noiseInSize,
+    size_t disIteration):
+    trainData(trainData),
+    initializeRule(initializeRule),
+    generator(generator),
+    discriminator(discriminator),
+    trainGenerator(false),
+    batchSize(batchSize),
+    noiseInSize(noiseInSize),
+    disIteration(2 * disIteration),
+    reset(false)
+{
+  discriminator.network.insert(discriminator.network.begin(),
+      new IdentityLayer<>());
+  trainGenerator = false;
+  iterations = 1;
+  offset = 0;
+  iterationDiscriminator = disIteration;
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Reset()
+{
+  std::cout << "Reset function" << std::endl;
+  size_t generatorWeights = 0;
+  size_t discriminatorWeights = 0;
+
+  for (size_t i = 0; i < generator.network.size(); ++i)
+    generatorWeights += boost::apply_visitor(weightSizeVisitor,
+        generator.network[i]);
+
+  assert(generatorWeights > 0);
+
+  for (size_t i = 0; i < discriminator.network.size(); ++i)
+    discriminatorWeights += boost::apply_visitor(weightSizeVisitor,
+        discriminator.network[i]);
+
+  assert(discriminatorWeights > 0);
+
+  parameter.set_size(generatorWeights + discriminatorWeights, 1);
```

Comment:
I tried the strategy that you stated above. The problem seems to be that both Network parameters ie Generator network and Discriminator network are initialized to the same value.